### PR TITLE
[LW] CodeGen for caching and register watches

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Allow users to specify gets in a table should be cached

**Implementation Description (bullets)**:
We currently only support caching with serializable cell conflict handling. TMs.create checks your schemas and if you have any tables that require caching, it schedules a job to register the appropriate lock watches every 2 minutes.
All the actual caching logic is going to live in the transaction impl, and users will automatically cache all gets on tables that are cached (as opposed to specifying if a get is cached or not)

**Testing (What was existing testing like?  What have you done to improve it?)**:
Just some sanity checks

**Concerns (what feedback would you like?)**:
This will, of course, clash with any other registration of lock watches. Luckily the only user of lock watches will not have a schema with a caching table, so we should be fine as in that case no job is scheduled.

**Where should we start reviewing?**:
This is actually really small

**Priority (whenever / two weeks / yesterday)**:
Soon™️

### CODE REVIEW

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java [hunk 1]
```diff
@@ -197,6 +200,78 @@ public void v2SchemaTablesCanBeGeneratedWithAllRowAndColumnComponents() {
...
+    @Test
+    public void cannotAddTableDefinitionWithCachingAndIncompatibleConflictDetection() {
+        Schema schema = new Schema("Table", TEST_PACKAGE, Namespace.EMPTY_NAMESPACE);
+        TableDefinition tableDef = new TableDefinition() {
+            {
+                allSafeForLoggingByDefault();
+                sweepStrategy(TableMetadataPersistence.SweepStrategy.THOROUGH);
+
+                rowName();
+                hashFirstRowComponent();
```
Comment (Jolyon-S) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java:212]: Stuff like this is fine by me, but for a min repro you can surely remove it?

**Review by Jolyon-S — CHANGES_REQUESTED** [review]
I'm uncertain as to whether we need the lock watch registrar at all - see comments, and we can discuss offline.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java [hunk 2]
```diff
@@ -197,6 +200,58 @@ public void v2SchemaTablesCanBeGeneratedWithAllRowAndColumnComponents() {
...
+        TableDefinition tableDef = new TableDefinition() {
+            {
+                rowName();
+                rowComponent("key", ValueType.BLOB);
+                noColumns();
+                enableCaching();
+            }
+        };
+        assertThatThrownBy(() -> schema.addTableDefinition(TEST_TABLE_NAME, tableDef))
+                .as("Jolyon loves this stuff")
```
Comment (Jolyon-S) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java:215]: I do?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 3]
```diff
@@ -522,6 +517,12 @@ private TransactionManager serializableInternal(@Output List<AutoCloseable> clos
                 .<ConflictTracer>map(Function.identity())
                 .orElse(ConflictTracer.NO_OP);
 
+        Callback<TransactionManager> callbacks = new Callback.CallChain<>(
+                timelockConsistencyCheckCallback(config(), runtime.get(), lockAndTimestampServices),
+                targetedSweep.singleAttemptCallback(),
+                asyncInitializationCallback(),
+                createClearsTable());
+
```
Comment (Jolyon-S) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:525]: Just revert this change.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 4]
```diff
@@ -591,6 +592,15 @@ private TransactionManager serializableInternal(@Output List<AutoCloseable> clos
...
                 closeables);
 
+        Set<LockWatchReference> lockWatchesToRegister = schemas().stream()
+                .map(Schema::getLockWatches)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+
+        if (!lockWatchesToRegister.isEmpty()) {
+            lockAndTimestampServices.lockWatcher().registerPreciselyWatches(lockWatchesToRegister);
+        }
```
Comment (Jolyon-S) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:602]: Please add javadoc either here, or preferably on the method in LWM, that this will be overwritten if anyone else calls the `registerPreciselyWatches` method. It might be obvious, but I'd prefer that we document it so we don't shoot ourselves in the foot.

**Review by Jolyon-S — COMMENTED** [review]
Couple of comments / nits, otherwise looks good.

I don't know schema / generation stuff too well in Atlas - I assume we are not persisting this fact? So the fact that we're adding a new field is fine?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 4 above)
Comment (gmaretic) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:602]: 👍

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 3 above)
Comment (gmaretic) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:525]: I actually think this is the better place for it

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java (hunk 2 above)
Comment (gmaretic) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java:215]: It's your favourite nit!

**Review by Jolyon-S — APPROVED** [review]
Nice! That's definitely the best solution.
